zpl2.py

```diff
@@ -425,3 +427,94 @@ def graphic_circle(self, right, down, graphic_format):
             stop=self._field_data_stop(),
         )
         self._write_command(command)
+
+    @staticmethod
+    def encode_grf_ascii(pil_image):
+        """Converts a PIL image to a .GRF file encoded in ascii for use
+        with zebra printers.
+        @:return array of strings, each string is a pixel row, containing
+        4 pixel per ascii character
+        """
+
+        def round_up(value, multiple):
+            result = value
+            rest = value % multiple
+            if rest:
+                result += multiple - rest
+            return result
+
+        def get_byte_val(bits, word_len):
+            if len(bits) == 0:
+                return 0
+            else:
+                byte_value = bits[0] << word_len - 1
+                return byte_value + get_byte_val(bits[1:], word_len - 1)
+
+        # make sure we have a monochrome image, and load it.
+        data = pil_image.convert("1").load()
+        sizex, sizey = pil_image.size
+
+        # increase row size to the nearest multiple of 8 pixels
+        # The length of the first row of pixels is set in bytes and must be
+        # a multiple of 8 bytes.
+        # The zebra tools also round the x size of images to the nearest of
+        # 8 pixels.
+        row_size = round_up(sizex, 8)
+        # construct an array y - x
+        # the first row of the array, is the first line of pixels in the
+        # image
+        # Fill all values with white
+        pixels = [[1 for x in range(0, row_size)] for y in range(0, sizey)]
+        for xpos in range(0, sizex):
+            for ypos in range(0, sizey):
+                input_value = data[xpos, ypos]
+                # PIL: pixel value 0 = white , pixel value 255 = black
+                # GRF: 1=white, 0=black
+                bit_value = input_value >> 7 ^ 1
```

Comment:
Hmmm I'm confused in this math. If I'm reading this right, `input_value` can be anything from 0 to 255? And then `bit_value` can be either 0 or 1 depending on this math here?

Assuming the above, I played with the math and can't find a difference between it and the below. Am I missing something, or is it possible to actually simplify like this?


```python
bit_value = input_value < 128
```

Reply:
On further thought, am I also understanding correctly that all of this logic is just to convert the image from color/greyscale into an actual black and white image? Why not just use PIL?

```python
from PIL import Image 
image_file = Image.open("convert_image.png") # open color image
image_file = image_file.convert('1') # convert image to black and white
```